### PR TITLE
tests for mock payment screen field validations

### PR DESCRIPTION
**Pull request by nelabhotlaR**
**The PR consists of:** 
1) Weather shopper Mock Payment screen field validations tests: **tests/test_weather_shopper_payment_app.py**

2) added code in core_helpers/mobile_app_helper.py - to get the source code of the page.

3) added code in core_helpers/selenium_action_objects.py - to hide the keyboard.

4) .circleci/config.yml - added lines to install tesseract.

5) In conf/locators_conf.py - # mock payment screen field names defined as a list.
                                             # used this list to navigate the cursor on the field. 
6) in conf/weather_shopper_mobile_conf.py added test data.

- 7) In page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py added three new methods 1) image_to_string() - extract text from image
- find_string() - to search substring from the text
- preprocess_image() - to enhance the image quality.

8) In requirements.txt - pytesseract package 

invalid-card

image-field
payment-validation
invalid-expiry-date

To run the test
============
`python3 -m pytest -s tests/test_weather_shopper_payment_app.py  --app_name weather_shopper.apk  --app_path /usr/local/android-studio/bin --app_package com.qxf2.weathershopper --remote_flag y`

### CODE REVIEW

File: core_helpers/mobile_app_helper.py [hunk 1]
```diff
@@ -403,3 +403,12 @@ def scroll_forward(self, distance=150):
     def start(self):
         "Dummy method to be over-written by child classes"
         pass
+    # to get source code of the app.
```
[the remark this answers is not in the document]
Comment (nelabhotlaR) [reply core_helpers/mobile_app_helper.py:406]: I Provided the space between methods. Comments started with capital letters.

File: page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py [hunk 2]
```diff
@@ -97,3 +100,99 @@ def submit_payment_details(self,card_type,email,card_number,card_expiry,card_cvv
         result_flag &= self.enter_card_cvv(card_cvv)
         result_flag &= self.submit_payment()
         return result_flag
+    """
```
[the remark this answers is not in the document]
Comment (nelabhotlaR) [reply page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py:103]: I modified the comments section.

File: page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py [hunk 3]
```diff
@@ -97,3 +100,99 @@ def submit_payment_details(self,card_type,email,card_number,card_expiry,card_cvv
...
+    navigate_to_field - Navigating the cursor to the field
+    and taking the screenshot of the error prompt. 
+    """
+    @Wrapit._exceptionHandler
+    def navigate_to_field(self, fieldname,screenshotname):
```
[the remark this answers is not in the document]
Comment (nelabhotlaR) [reply page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py:108]: I Modified the logic. Check and let me know

File: core_helpers/selenium_action_objects.py [hunk 4]
```diff
@@ -297,3 +297,8 @@ def teardown(self):
         "Tears down the driver"
         self.driver.quit()
         self.reset()
+    
+    # to minimise the keyboard on the screen.
+    def hide_keyboard(self):
+        # To minimise the keyboard.
+        self.driver.hide_keyboard()
```
[the remark this answers is not in the document]
Comment (avinash010) [reply core_helpers/selenium_action_objects.py:303]: @nelabhotlaR Look for other methods like scroll down. We need to have a try catch and also return a result Flag which says True or False

File: page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py [hunk 5]
```diff
@@ -97,3 +101,82 @@ def submit_payment_details(self,card_type,email,card_number,card_expiry,card_cvv
...
+
+    @Wrapit._exceptionHandler
+    def find_string(self,text,substring):
+        """
+        Check if the substring is in the input string
+        """
+        if substring in text:
+            print(f'Substring:{substring}')
+            return True
+
```
Comment (avinash010) [inline page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py:163]: The image_to_string should be get_string_from_image which should return a string if function works fine or should return null.
This method should be called from test. We can call this as validate_string. This should take in the text and substring and do the validation. Return False if the string match fails.

File: core_helpers/selenium_action_objects.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nelabhotlaR) [reply core_helpers/selenium_action_objects.py:303]: Implemented the suggestion

File: conf/locators_conf.py [hunk 6]
```diff
@@ -74,14 +74,18 @@
 checkbox = "xpath,//android.widget.TextView[@text='{}']/ancestor::android.widget.LinearLayout/preceding-sibling::android.widget.CheckBox"
 delete_from_cart_button = "id,deleteSelectedButton"
 checkout_button = "id,checkoutButton"
-payment_method_dropdown = "xpath,//android.widget.TextView[@text='Select Payment Method']"
+#payment_method_dropdown = "xpath,//android.widget.TextView[@text='Select Payment Method']"
```
[the remark this answers is not in the document]
Comment (nelabhotlaR) [reply conf/locators_conf.py:77]: The older one was deleted.

File: conf/weather_shopper_mobile_conf.py [hunk 7]
```diff
@@ -4,4 +4,56 @@
...
+"card_type": "Credit Card",
+"email": "qxf2tester@example.com",
+"card_number": "1234567890",
+"card_expiry": "12/25",
+"card_cvv": "123",
+"image_path": "./screenshots/test_weather_shopper_payment_app/navigate_to_cardnumber_field.png",
+"substring": "Invalid card number"    
+}
+
+invalid_expirydate_in_payment_details = {
```
[the remark this answers is not in the document]
Comment (nelabhotlaR) [reply conf/weather_shopper_mobile_conf.py:31]: Removed.

File: page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py [hunk 8]
```diff
@@ -97,3 +101,82 @@ def submit_payment_details(self,card_type,email,card_number,card_expiry,card_cvv
...
+            locators.fieldnames[2]: [locators.payment_email, locators.payment_card_expiry],
+            locators.fieldnames[3]: [locators.payment_email, locators.payment_card_cvv]
+        }
+        actions = field_actions.get(
+            fieldname, [locators.payment_email, locators.payment_card_expiry]
+        )
+        # Perform the actions
+        for action in actions:
+            result_flag = self.click_element(action)
+
```
[the remark this answers is not in the document]
Comment (nelabhotlaR) [reply page_objects/examples/weather_shopper_mobile_app/weather_shopper_payment_page.py:124]: if fieldname == "email":
            result_flag = self.click_element(locators.payment_email)
            self.hide_keyboard()
            self.save_screenshot(screenshot_name=screenshotname)
        elif fieldname == "card number":
            result_flag = self.click_element(locators.payment_card_number)
            self.hide_keyboard()
            self.save_screenshot(screenshot_name=screenshotname)
        elif fieldname == "date":
            self.click_element(locators.payment_email)
            result_flag = self.click_element(locators.payment_card_expiry)
            self.hide_keyboard()
            self.save_screenshot(screenshot_name=screenshotname)
        elif fieldname == "cvv":
            self.click_element(locators.payment_email)
            result_flag = self.click_element(locators.payment_card_cvv)
            self.hide_keyboard()
            self.save_screenshot(screenshot_name=screenshotname)
        else:
            self.click_element(locators.payment_email)
            result_flag = self.click_element(locators.payment_card_expiry)
            self.hide_keyboard()
            self.save_screenshot(screenshot_name=screenshotname)
        return result_flag

**Pull request merged**